command/cmd_delete.go

```diff
@@ -97,3 +100,89 @@ func Delete(
 
 	return merror
 }
+
+// expandSources is a non-blocking argument dispatcher.
+// It creates a object channel by walking and expanding the given source urls.
+// If the url has a glob, it creates a goroutine to list storage items and sends them to
+// object channel, otherwise it creates storage object from the original source.
+func expandSources(
+	ctx context.Context,
+	client storage.Storage,
+	srcurls ...*url.URL,
+) <-chan *storage.Object {
+	objChan := make(chan *storage.Object)
+	go func() {
+		defer close(objChan)
+
+		var wg sync.WaitGroup
+		var objFound bool
+
+		for _, origSrc := range srcurls {
+			// call storage.List for only walking operations.
+			if origSrc.HasGlob() {
```

Comment:
what about directories?

Reply:
It uses expandSource() method now. It makes stat call and walks directories if the source URL is local. I also added e2e tests for the directory usage.